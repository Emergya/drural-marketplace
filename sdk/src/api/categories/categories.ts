import ApolloClient from "apollo-client";
import { CategoryAncestorsListVariables } from "../../queries/gqlTypes/CategoryAncestorsList";
import { CategoryChildrenListVariables } from "../../queries/gqlTypes/CategoryChildrenList";
import {
  CategoryDetails as CategoryDetailsQuery,
  CategoryDetailsVariables,
} from "../../queries/gqlTypes/CategoryDetails";
import { CategoryDetails as CategoryDetailsFragment } from "../../fragments/gqlTypes/CategoryDetails";
import {
  CategoryList as CategoryListQuery,
  CategoryListVariables,
} from "../../queries/gqlTypes/CategoryList";
import { BaseCategory } from "../../fragments/gqlTypes/BaseCategory";
import { WithDetails, WithList } from "../types";
import { CategoryList } from "./CategoryList";
import { CategoryAncestorsList } from "./CategoryAncestorsList";
import { CategoryChildrenList } from "./CategoryChildrenList";
import { CategoryDetails } from "./CategoryDetails";

export class CategoriesAPI
  implements
    WithDetails<
      CategoryDetailsQuery,
      CategoryDetailsFragment,
      CategoryDetailsVariables
    >,
    WithList<CategoryListQuery, BaseCategory, CategoryListVariables> {
  private client: ApolloClient<any>;

  constructor(client: ApolloClient<any>) {
    this.client = client;
  }

  /**
   * Method returning category details
   * @param variables Details parameters
   */
  getDetails = async (variables: CategoryDetailsVariables) => {
    const details = new CategoryDetails(this.client);

    await details.init(variables);

    return details;
  };

  /**
   * Method returning list of categories with ability to request next page
   * @param params List parameters
   */
  getList = async (variables: CategoryListVariables) => {
    const list = new CategoryList(this.client);

    await list.init(variables);

    return list;
  };

  getAncestors = async (variables: CategoryAncestorsListVariables) => {
    const list = new CategoryAncestorsList(this.client);

    await list.init(variables);

    return list;
  };

  getChildren = async (variables: CategoryChildrenListVariables) => {
    const list = new CategoryChildrenList(this.client);

    await list.init(variables);

    return list;
  };
}
