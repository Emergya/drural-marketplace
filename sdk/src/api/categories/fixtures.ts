export const categoryWithChildren = "Q2F0ZWdvcnk6Nw==";
export const categoryWithParent = "Q2F0ZWdvcnk6MjI=";
