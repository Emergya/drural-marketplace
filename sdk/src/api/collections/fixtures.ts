export const collectionId = "Q29sbGVjdGlvbjox";
export const collectionSlug = "summer-collection";
