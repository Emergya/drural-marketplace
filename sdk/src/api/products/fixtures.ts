export const productId = "UHJvZHVjdDo4NA==";
export const productSlug = "seaman-beer";
