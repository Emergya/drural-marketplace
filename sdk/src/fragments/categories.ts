import gql from "graphql-tag";

export const baseCategoryFragment = gql`
  fragment BaseCategory on Category {
    id
    name
    slug
    seoDescription
    seoTitle
  }
`;

export const categoryFragment = gql`
  ${baseCategoryFragment}
  fragment CategoryDetails on Category {
    ...BaseCategory
    description
  }
`;
