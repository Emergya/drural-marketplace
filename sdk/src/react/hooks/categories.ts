import { CategoryDetails } from "../../api/categories/CategoryDetails";
import { CategoryAncestorsList } from "../../api/categories/CategoryAncestorsList";
import { CategoryChildrenList } from "../../api/categories/CategoryChildrenList";
import { CategoryList } from "../../api/categories/CategoryList";
import { makeDetails, makeList } from "./utils";

export const useCategoryList = makeList(client => new CategoryList(client));
export const useCategoryAncestorsList = makeList(
  client => new CategoryAncestorsList(client)
);
export const useCategoryChildrenList = makeList(
  client => new CategoryChildrenList(client)
);
export const useCategoryDetails = makeDetails(
  client => new CategoryDetails(client)
);
