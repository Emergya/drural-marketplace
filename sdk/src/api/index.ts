import ApolloClient from "apollo-client";

import { LocalStorageManager } from "../data";
import { ApolloClientManager } from "../data/ApolloClientManager";
import { LocalStorageHandler } from "../helpers/LocalStorageHandler";
import { JobsManager } from "../jobs";
import { SaleorState } from "../state";
import { Config } from "../types";
import { AuthAPI } from "./Auth";
import APIProxy from "./APIProxy";
import { SaleorCartAPI } from "./Cart";
import { SaleorCheckoutAPI } from "./Checkout";
import { CollectionsAPI } from "./collections/collections";
import { CategoriesAPI } from "./categories/categories";
import { ProductsAPI } from "./products/products";

export * from "./Checkout";
export * from "./Cart";

export class SaleorAPI {
  auth: AuthAPI;

  checkout: SaleorCheckoutAPI;

  cart: SaleorCartAPI;

  categories: CategoriesAPI;

  collections: CollectionsAPI;

  products: ProductsAPI;

  /**
   * @deprecated Please do not use it anymore. Reference to API Proxy will be removed in future.
   * Now it just exists for legacy React hooks, which also will be removed.
   */
  legacyAPIProxy: APIProxy;

  constructor(
    client: ApolloClient<any>,
    apiProxy: APIProxy,
    config: Config,
    onStateUpdate?: () => any
  ) {
    this.legacyAPIProxy = apiProxy;

    const localStorageHandler = new LocalStorageHandler();
    const apolloClientManager = new ApolloClientManager(client);
    const jobsManager = new JobsManager(
      localStorageHandler,
      apolloClientManager
    );
    const saleorState = new SaleorState(
      config,
      localStorageHandler,
      apolloClientManager,
      jobsManager
    );
    const localStorageManager = new LocalStorageManager(
      localStorageHandler,
      saleorState
    );

    if (onStateUpdate) {
      saleorState.subscribeToNotifiedChanges(onStateUpdate);
    }

    this.auth = new AuthAPI(saleorState, jobsManager, config);
    this.checkout = new SaleorCheckoutAPI(saleorState, jobsManager, config);
    this.cart = new SaleorCartAPI(
      localStorageManager,
      apolloClientManager,
      saleorState,
      jobsManager,
      config
    );
    this.categories = new CategoriesAPI(client);
    this.collections = new CollectionsAPI(client, config);
    this.products = new ProductsAPI(client, config);

    this.legacyAPIProxy.attachAuthListener(authenticated => {
      if (!authenticated) {
        localStorageHandler.setCheckout({});
        localStorageHandler.setPayment({});
        localStorageHandler.setJobs(null);
      }
    });
  }
}
