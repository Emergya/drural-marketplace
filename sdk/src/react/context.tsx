import React from "react";

import { SaleorAPI } from "../api";
import { ConfigInput } from "../types";

export type SaleorContextType = { api: SaleorAPI; config: ConfigInput };

export const SaleorContext = React.createContext<SaleorContextType | null>(
  null
);
