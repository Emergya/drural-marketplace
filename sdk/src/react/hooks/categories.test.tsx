import { act, renderHook } from "@testing-library/react-hooks";
import React from "react";
import { ApolloProvider } from "react-apollo";
import { SaleorContext } from "../context";
import { setupContextAndAPI } from "../../../testUtils/context";
import { setupRecording } from "../../../testUtils/api";
import {
  useCategoryList,
  useCategoryChildrenList,
  useCategoryAncestorsList,
  useCategoryDetails,
} from "./categories";
import * as fixtures from "../../api/categories/fixtures";

setupRecording();

describe("useCategoryList", () => {
  let wrapper: React.FC<{}>;

  beforeAll(async () => {
    const { client, context } = await setupContextAndAPI();

    wrapper = ({ children }) => (
      <SaleorContext.Provider value={context}>
        <ApolloProvider client={client}>{children}</ApolloProvider>
      </SaleorContext.Provider>
    );
  });

  it("can fetch category", async () => {
    const { result } = renderHook(
      () =>
        useCategoryDetails({
          id: fixtures.categoryWithChildren,
        }),
      {
        wrapper,
      }
    );

    expect(result.current.data).toBe(undefined);
    expect(result.current.loading).toBe(true);

    // @ts-ignore
    await act(() => result.current.current);

    expect(result.current.data).toMatchSnapshot();
    expect(result.current.loading).toBe(false);
  });

  it("can fetch categories", async () => {
    const { result } = renderHook(
      () =>
        useCategoryList({
          first: 20,
        }),
      {
        wrapper,
      }
    );

    expect(result.current.data).toBe(undefined);
    expect(result.current.loading).toBe(true);

    // @ts-ignore
    await act(() => result.current.current);

    expect(result.current.data).toMatchSnapshot();
    expect(result.current.loading).toBe(false);
  });

  it("can fetch next page", async () => {
    const { result } = renderHook(
      () =>
        useCategoryList({
          first: 1,
        }),
      {
        wrapper,
      }
    );

    expect(result.current.data).toBe(undefined);
    expect(result.current.loading).toBe(true);

    // @ts-ignore
    await act(() => result.current.current);

    expect(result.current.data).toMatchSnapshot();
    expect(result.current.loading).toBe(false);

    await act(result.current.next);

    expect(result.current.data).toMatchSnapshot();
    expect(result.current.loading).toBe(false);
  });
});

describe("useCategoryAncestorsList", () => {
  let wrapper: React.FC<{}>;

  beforeAll(async () => {
    const { client, context } = await setupContextAndAPI();

    wrapper = ({ children }) => (
      <SaleorContext.Provider value={context}>
        <ApolloProvider client={client}>{children}</ApolloProvider>
      </SaleorContext.Provider>
    );
  });

  it("can fetch category ancestors", async () => {
    const { result } = renderHook(
      () =>
        useCategoryAncestorsList({
          first: 20,
          id: fixtures.categoryWithParent,
        }),
      {
        wrapper,
      }
    );

    expect(result.current.data).toBe(undefined);
    expect(result.current.loading).toBe(true);

    // @ts-ignore
    await act(() => result.current.current);

    expect(result.current.data).toMatchSnapshot();
    expect(result.current.loading).toBe(false);
  });
});

describe("useCategoryChildrenList", () => {
  let wrapper: React.FC<{}>;

  beforeAll(async () => {
    const { client, context } = await setupContextAndAPI();

    wrapper = ({ children }) => (
      <SaleorContext.Provider value={context}>
        <ApolloProvider client={client}>{children}</ApolloProvider>
      </SaleorContext.Provider>
    );
  });

  it("can fetch subcategories", async () => {
    const { result } = renderHook(
      () =>
        useCategoryChildrenList({
          first: 20,
          id: fixtures.categoryWithChildren,
        }),
      {
        wrapper,
      }
    );

    expect(result.current.data).toBe(undefined);
    expect(result.current.loading).toBe(true);

    // @ts-ignore
    await act(() => result.current.current);

    expect(result.current.data).toMatchSnapshot();
    expect(result.current.loading).toBe(false);
  });

  it("can fetch next page", async () => {
    const { result } = renderHook(
      () =>
        useCategoryChildrenList({
          first: 1,
          id: fixtures.categoryWithChildren,
        }),
      {
        wrapper,
      }
    );

    expect(result.current.data).toBe(undefined);
    expect(result.current.loading).toBe(true);

    // @ts-ignore
    await act(() => result.current.current);

    expect(result.current.data).toMatchSnapshot();
    expect(result.current.loading).toBe(false);

    await act(result.current.next);

    expect(result.current.data).toMatchSnapshot();
    expect(result.current.loading).toBe(false);
  });
});
