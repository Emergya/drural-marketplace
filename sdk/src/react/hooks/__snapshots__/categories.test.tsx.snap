// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useCategoryAncestorsList can fetch category ancestors 1`] = `
Array [
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6Nw==",
    "name": "Accessories",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "accessories",
  },
]
`;

exports[`useCategoryChildrenList can fetch next page 1`] = `
Array [
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MjA=",
    "name": "Paints",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "paints",
  },
]
`;

exports[`useCategoryChildrenList can fetch next page 2`] = `
Array [
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MjA=",
    "name": "Paints",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "paints",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MjI=",
    "name": "Homewares",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "homewares",
  },
]
`;

exports[`useCategoryChildrenList can fetch subcategories 1`] = `
Array [
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MjA=",
    "name": "Paints",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "paints",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MjI=",
    "name": "Homewares",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "homewares",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MjM=",
    "name": "Audiobooks",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "ebooks",
  },
]
`;

exports[`useCategoryList can fetch categories 1`] = `
Array [
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6Nw==",
    "name": "Accessories",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "accessories",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6OA==",
    "name": "Groceries",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "groceries",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6OQ==",
    "name": "Apparel",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "apparel",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MTA=",
    "name": "T-shirts",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "t-shirts",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MTE=",
    "name": "Polo Shirts",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "polo-shirts",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MTI=",
    "name": "Hoodies",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "hoodies",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MTM=",
    "name": "Footwear",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "footwear",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MTQ=",
    "name": "Juices",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "juices",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MTU=",
    "name": "Alcohol",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "alcohol",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MjA=",
    "name": "Paints",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "paints",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MjI=",
    "name": "Homewares",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "homewares",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6MjM=",
    "name": "Audiobooks",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "ebooks",
  },
]
`;

exports[`useCategoryList can fetch category 1`] = `
Object {
  "__typename": "Category",
  "backgroundImage": Object {
    "__typename": "Image",
    "alt": "",
    "url": "https://master.staging.saleor.cloud/media/category-backgrounds/accessories.jpg",
  },
  "description": "{\\"blocks\\": [{\\"data\\": {\\"text\\": \\"\\"}, \\"type\\": \\"paragraph\\"}]}",
  "id": "Q2F0ZWdvcnk6Nw==",
  "name": "Accessories",
  "seoDescription": "",
  "seoTitle": "",
  "slug": "accessories",
}
`;

exports[`useCategoryList can fetch next page 1`] = `
Array [
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6Nw==",
    "name": "Accessories",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "accessories",
  },
]
`;

exports[`useCategoryList can fetch next page 2`] = `
Array [
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6Nw==",
    "name": "Accessories",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "accessories",
  },
  Object {
    "__typename": "Category",
    "id": "Q2F0ZWdvcnk6OA==",
    "name": "Groceries",
    "seoDescription": "",
    "seoTitle": "",
    "slug": "groceries",
  },
]
`;
